### PR TITLE
feat(client|redux): convert recommendations area to ts

### PR DESCRIPTION
**Pull request by danielbento92**
## Description

- Conversion to typescript:
 - Client/recommendations
 - Redux/recommendations

## Checklist

- [x] The commit message follows our guidelines
- [x] Tests for the respective changes have been added
- [x] The code is commented, particularly in hard-to-understand areas
- [x] The labels and/or milestones were added

## Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

File: packages/redux/src/recommendations/__tests__/reducer.test.ts [hunk 1]
```diff
@@ -1,11 +1,13 @@
-import { mockRecommendationsStrategy } from '../../../tests/__fixtures__/initialReduxState';
+import { INITIAL_STATE } from '../reducer';
+import { mockRecommendationsStrategy } from 'tests/__fixtures__/recommendations';
 import reducer, { actionTypes } from '..';
+import type { State } from '../types';
 
 let initialState;
```
Comment (boliveira) [inline packages/redux/src/recommendations/__tests__/reducer.test.ts:6]: This has `any` type, so typescript cannot know if the properties you are accessing from initialState are there or not. Can you add a type to initialState?

File: packages/redux/src/recommendations/__tests__/reducer.test.ts [hunk 2]
```diff
@@ -48,19 +50,21 @@ describe('recommendations reducer', () => {
     });
 
     it('should handle other actions by returning the previous state', () => {
-      const state = {
+      const state: State = {
         error: {
           [mockRecommendationsStrategy]: error,
```
Comment (boliveira) [inline packages/redux/src/recommendations/__tests__/reducer.test.ts:55]: Typescript is complaining that `error` is a string and not an `Error` instance. Can you change it to an Error, please?

File: packages/redux/src/recommendations/types/state.types.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import type { CombinedState } from 'redux';
+import type { Error } from '@farfetch/blackout-client/types';
+import type { ProductRecommendation } from '@farfetch/blackout-client/recommendation/types';
+
+export type State = CombinedState<{
+  error: Record<string, Error | null | undefined>;
+  isLoading: Record<string, boolean>;
+  result: ProductRecommendationResult;
+}>;
```
Comment (boliveira) [inline packages/redux/src/recommendations/types/state.types.ts:9]: State can be undefined as well as the recommendations slice is defined in StoreState.types.ts as being optional. Can you change this, please? Make sure then in all selectors that use a value of this type to handle the undefined case as well.

File: packages/client/src/recommendations/getProductRecommendations.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * @param query - Query object with product recommendation parameters to apply.
+ * @param config - Custom configurations to send to the client
+ * instance (axios).
+ *
+ * @returns Promise object.
+ */
+
+const getProductRecommendations: GetProductRecommendations = (
+  query,
```
Comment (boliveira) [inline packages/client/src/recommendations/getProductRecommendations.ts:17]: This is a breaking change, however the commit message does not contain the BREAKING MESSAGE description containing the necessary changes to migrate to this new version. Also do not forget to update the [migration guide in Confluence](https://gitlab.fftech.info/fps/platform/box/wikis/@fps-core/Migration-guide-from-v1.0.0-to-v2.0.0) with the same information.

File: packages/redux/src/recommendations/__tests__/reducer.test.ts [hunk 5]
```diff
@@ -1,11 +1,13 @@
...
 import reducer, { actionTypes } from '..';
+import type { State } from '../types';
 
 let initialState;
-
+const mockAction = { type: 'this_is_a_mock_action' };
 describe('recommendations reducer', () => {
   beforeEach(() => {
-    initialState = reducer();
+    initialState = reducer(INITIAL_STATE, mockAction);
```
Comment (boliveira) [inline packages/redux/src/recommendations/__tests__/reducer.test.ts:10]: INITIAL_STATE should be undefined to make sure the reducer is initialised with the INITIAL_STATE.

File: packages/redux/src/recommendations/__tests__/reducer.test.ts [hunk 6]
```diff
@@ -48,19 +50,21 @@ describe('recommendations reducer', () => {
...
 
-      expect(reducer(state).error).toEqual(state.error);
+      expect(reducer(state, mockAction).error).toEqual(state.error);
     });
   });
 
   describe('isLoading() reducer', () => {
     it('should return the initial state', () => {
-      const state = reducer();
+      const state = reducer(initialState, mockAction);
```
Comment (boliveira) [inline packages/redux/src/recommendations/__tests__/reducer.test.ts:67]: Same as above, initialState should be replaced with `undefined`. There are a few places in this file that contain the same pattern, can you review all of them and change where necessary?

File: packages/client/src/recommendations/types/getProductRecommendations.types.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  countryCode?: string;
+  gender?: string;
+  userType?: string;
+  userIdentifier?: string;
+  categoryId?: number;
+  brandId?: string;
+  merchantId?: number;
+};
+
+export type GetProductRecommendation = {
```
Comment (boliveira) [inline packages/client/src/recommendations/types/getProductRecommendations.types.ts:21]: This should be called ProductRecommendation as it is independent from the request method.

File: packages/client/src/recommendations/types/getProductRecommendations.types.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  Internal,
+  Sponsored,
+}
+
+export type GetRecommendationTrackerItem = {
+  type: GetRecommendationTrackerType;
+  targetUri: string;
+};
+
+export type GetProductItem = {
```
Comment (boliveira) [inline packages/client/src/recommendations/types/getProductRecommendations.types.ts:43]: Same as above, this should be called ProductItem and all other types that are not specific to the get request should not have the request method prefixed to the type.

Comment (boliveira) [reply packages/client/src/recommendations/types/getProductRecommendations.types.ts:43]: Actually never mind, I just noticed you are doing that already.

File: packages/redux/src/recommendations/reducer.ts [hunk 9]
```diff
@@ -73,23 +69,38 @@ const result = (
...
+ * @returns New state.
  */
-export default combineReducers({
-  error,
-  isLoading,
-  result,
-});
+const recommendationsReducer: ReducerSwitch<State, AnyAction> = (
+  state,
+  action = { type: undefined },
```
Comment (boliveira) [inline packages/redux/src/recommendations/reducer.ts:97]: Can you review this default parameter value to match the other reducers (forms namely)?

File: packages/redux/src/recommendations/selectors.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * @param  state - Application state.
+ * @param  strategyName - The strategy name for the specific set of recommendations.
+ *
+ * @returns The result object containing the product recommendations.
+ */
+export const getProductRecommendationsByStrategyName = (
+  state: StoreState,
+  strategyName: string,
+): unknown | undefined => get(getProductRecommendations(state), strategyName);
```
Comment (boliveira) [inline packages/redux/src/recommendations/selectors.ts:73]: In this case it is preferable to return unknown only as in practice the user will have to check the value anyway.

File: packages/redux/src/recommendations/__tests__/reducer.test.ts (hunk 1 above)
Comment (danielbento92) [reply packages/redux/src/recommendations/__tests__/reducer.test.ts]: Source updated, can you check again.

File: packages/redux/src/recommendations/__tests__/reducer.test.ts (hunk 2 above)
Comment (danielbento92) [reply packages/redux/src/recommendations/__tests__/reducer.test.ts]: Source updated, can you check again.

File: packages/client/src/recommendations/getProductRecommendations.ts (hunk 4 above)
Comment (danielbento92) [reply packages/client/src/recommendations/getProductRecommendations.ts:17]: Added Breaking change label to pull request.

File: packages/redux/src/recommendations/__tests__/reducer.test.ts (hunk 5 above)
Comment (danielbento92) [reply packages/redux/src/recommendations/__tests__/reducer.test.ts]: Source updated, can you check again.

File: packages/redux/src/recommendations/__tests__/reducer.test.ts (hunk 6 above)
Comment (danielbento92) [reply packages/redux/src/recommendations/__tests__/reducer.test.ts]: Source updated, can you check again.

File: packages/redux/src/recommendations/__tests__/selectors.test.ts [hunk 11]
```diff
@@ -58,7 +58,7 @@ describe('Recommendations redux selectors', () => {
   });
 
   describe('getProductRecommendationsByStrategyName()', () => {
-    it('should get the result by strategy name', () => {
+    fit('should get the result by strategy name', () => {
```
[the remark this answers is not in the document]
Comment (danielbento92) [reply packages/redux/src/recommendations/__tests__/selectors.test.ts]: Source updated, can you check again.

File: packages/redux/src/recommendations/reducer.ts (hunk 9 above)
Comment (danielbento92) [reply packages/redux/src/recommendations/reducer.ts]: Source updated, can you check again.
FormReducer, updated too.

File: packages/redux/src/recommendations/selectors.ts (hunk 10 above)
Comment (danielbento92) [reply packages/redux/src/recommendations/selectors.ts]: Source updated, can you check again.

**Review by boliveira — APPROVED** [review]

**Review by ruifcnunes — APPROVED** [review]

**Pull request merged**